os/arch/xtensa/src/esp32/esp32_allocateheap.c

```diff
@@ -98,17 +98,32 @@ void up_allocate_heap(FAR void **heap_start, size_t *heap_size)
 	*heap_size = (size_t)((uintptr_t)&_eheap - (uintptr_t)&_sheap);
 }
 
+#if CONFIG_MM_REGIONS > 1
+
 /****************************************************************************
- * Name: xtensa_add_region
- *
- * Description:
- *   Memory may be added in non-contiguous chunks.  Additional chunks are
- *   added by calling this function.
- *
+ * Name: up_addregion
  ****************************************************************************/
+void up_addregion(void)
+{
 
-#if CONFIG_MM_REGIONS > 1
-void xtensa_add_region(void)
+}
+
+/****************************************************************************
+ * Name: up_addregion after spiram initialized
+ ****************************************************************************/
+void xtensa_up_addregion(void)
 {
+	int region_cnt;
+
+	for (region_cnt = 1; region_cnt < CONFIG_MM_REGIONS; region_cnt++) {
+		if (heapx_is_init[regionx_heap_idx[region_cnt]] != true) {
+			mm_initialize(&g_mmheap[regionx_heap_idx[region_cnt]], regionx_start[region_cnt], regionx_size[region_cnt]);
+			heapx_is_init[regionx_heap_idx[region_cnt]] = true;
+			continue;
+		}
+		mm_addregion(&g_mmheap[regionx_heap_idx[region_cnt]], regionx_start[region_cnt], regionx_size[region_cnt]);
+	}
```

Comment:
why need continue,  if the region is checked at first time ,  then not call mm_addregion?

Reply:
mm_initialize include the mm_addregion function call, so it could continue here